# Лабораторная работа №3

**Цель работы:** практическое применение библиотеки символьных вычислений Python
(SymPy) для решения задач дискретной математики и математической логики.

**Задачи:** 
1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и
расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим
выражением.

## 1. Решение диофантова уравнения

In [5]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
x, y = symbols("x, y", integer=True)
print(diophantine(3281*x + 4301*y + 34))

{(253*t_0 + 194, -193*t_0 - 148)}


Как видно из результата выше, решение диофантова уравнения 3281x + 4301y = -34 имеет вид: x = 253t + 194; y= -193t - 148

## 2. Расшифровка кодового слова

**Задание:** Русское слово из четырех вукв закодировано при помощи алгоритма RSA открытым ключом (e=11, m=33). Шифрованое сообение имеет вид (25;2;18;28). Подберите закрытую часть ключа и прочитайте исходное слово.

In [25]:
from sympy.crypto.crypto import rsa_private_key, decipher_rsa
alphabet = ['а', 'б', 'в', 'г', 'д', 'е', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о', 'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц',
            'ч', 'ш', 'щ', 'ы', 'ь', 'э','ю','я']
private_key = rsa_private_key(3, 11, 11)
f = alphabet[decipher_rsa(25, private_key) - 2] 
s = alphabet[decipher_rsa(2, private_key)-2]
t = alphabet[decipher_rsa(18, private_key) - 2]
four = alphabet[decipher_rsa(28, private_key)-2]
print('Закодированное слово: '+ f + s + t + four)

Закодированное слово: чары


## 3. Запись в формате ДНФ и КНФ булевой функции

f(x,y,z) = (z xor (yx)) or ((x xor y)(z or x))

In [34]:
from sympy.logic.boolalg import to_dnf, to_cnf
from sympy.abc import x, y, z
from sympy.logic.boolalg import Xor
print(to_dnf(Xor(z, (y & x)) | (Xor(x, y) & (z | x)), simplify=True))
(to_cnf(Xor(z, (y & x)) | (Xor(x, y) & (z | x)), simplify=True))

(x & ~z) | (z & ~x) | (z & ~y)


(x | z) & (~x | ~y | ~z)

## Вывод

В результате выполнения данной лабораторной работы была изучена бибилиотека sympy, а также были применены некоторые функции этой библиотеки для решения задач дискретной математики и математической логики.